test/frontend/frontend.test.js

```diff
@@ -0,0 +1,353 @@
+      window.XMLHttpRequest.prototype.originalXHR.restore();
+    }
+  }
+  
+  const cb = function(done) {
+    if (file === undefined ||
+        encryptedIV === undefined || 
+        fileHash === undefined || 
+        secretKey === undefined) {
+      assert.fail('Please run file sending tests before trying to receive the files.');
+      done();
+    }
+
+    FakeXHR.setup();
+    done();
+  }
+
+  before(cb)
+
+  after(function() {
+    FakeXHR.restore();
+  })
+
+  it('Should decrypt properly', function() {
+    const fr = new FileReceiver();
+    location.hash = secretKey;
+    return fr.download().then(([decrypted, name]) => {
+      assert(name);
+      assert(new Uint8Array(decrypted).toString() ===
+                  new Uint8Array(originalBlob).toString())
+    }).catch(err => {
+      console.log(err, err.stack);
+      assert.fail();
+    })
+  })
+
+  it('Should emit decrypting events', function() {
+    const fr = new FileReceiver();
+    location.hash = secretKey;
+
+    let testDecrypting = true;
+
+    fr.on('decrypting', isStillDecrypting => {
+      assert(!(!testDecrypting && isStillDecrypting));
+      testDecrypting = isStillDecrypting;
+    });
+
+    fr.on('safe', isSafe => {
+      assert(isSafe);
+    })
+
+    return fr.download().then(([decrypted, name]) => {
+      assert(decrypted);
+      assert(name);
+      assert(!testDecrypting);
+    }).catch(err => {
+      console.log(err, err.stack);
+      assert.fail();
+    })
+  })
+
+  it('Should emit hashing events', function() {
+    const fr = new FileReceiver();
+    location.hash = secretKey;
+
+    let testHashing = true;
+
+    fr.on('hashing', isStillHashing => {
+      assert(!(!testHashing && isStillHashing));
+      testHashing = isStillHashing;
+    });
+
+    fr.on('safe', isSafe => {
+      assert(isSafe);
+    })
+
+    return fr.download().then(([decrypted, name]) => {
+      assert(decrypted);
+      assert(name);
+      assert(!testHashing);
+    }).catch(err => {
+      assert.fail();
+    })
+  })
+
+  it('Should catch tampered checksums', function(done) {
+    // use the secret key and file hash of the previous file to encrypt,
+    // which has a different hash than this one (different strings)
+    const newFile = new FakeFile('hello_world.txt', 
+                      ['This is some data, with a changed hash.'])
+    const readRaw = new FileReader();
+    
+    readRaw.onload = function(event) {
+      const plaintext = new Uint8Array(this.result);
+      window.crypto.subtle.importKey(
+        'jwk',
+        {
+          kty: 'oct',
+          k: secretKey,
+          alg: 'A128GCM',
+          ext: true
+        },
+        {
+          name: 'AES-GCM'
+        },
+        true,
+        ['encrypt', 'decrypt']
+      )
+      .then(key => {
+        return window.crypto.subtle.encrypt(
+          {
+            name: 'AES-GCM',
+            iv: hexToArray(encryptedIV),
+            additionalData: fileHash,
```

Comment:
Add a comment here, this is the meat of this whole test, lets make it obvious that `fileHash` is the phony checksum